### PR TITLE
feat(deno): add `sync with meilisearch template`

### LINKED ISSUE #189: 🚀 Feature: Add Sync with Meilisearch for Deno
Appwrite Functions allow you to extend and customize your Appwrite server functionality by executing your custom code. 🤩 You can learn more at our official [Appwrite Functions docs](https://appwrite.io/docs/functions).

Your task is to implement the Sync with Meilisearch function in Deno. You can look at the existing [Appwrite Functions Templates](https://github.com/appwrite/templates) in another coding language.

Your function should behave and be implemented similarly to existing Sync with Meilisearch Node.js template.

### Tasks summary:

- [ ] Fork & clone Appwrite's [templates](https://github.com/appwrite/templates) repository
- [ ] Create a branch `feat-implement-sync-with-meilisearch-deno`
- [ ] Write code for the function in folder `deno/sync-with-meilisearch/`. 
- [ ] Test the function using Appwrite, **provide a video**.
- [ ] Write a `README.md` explaining what the function does, how to use it, and customize it. Here is [an example](https://github.com/appwrite/templates/tree/main/node/sync-with-meilisearch).
- [ ] Submit a pull request to our [templates](https://github.com/appwrite/templates) repository

If you need any help, contact us on our [Discord server](https://discord.gg/GSeTUeA).

Are you ready to work on this issue? 🤔 Let us know, and we will assign it to you 😊

**Happy Appwriting!**

### PR DESCRIPTION
**Pull request by Aunali321**
closes #189

### CODE REVIEW

**Review by loks0n — CHANGES_REQUESTED** [review]
Thanks for the contribution!

Please can you provide a test video as outline in https://github.com/appwrite/templates/issues/189

File: deno/sync-with-meilisearch/src/main.js
Comment (loks0n) [inline deno/sync-with-meilisearch/src/main.js:1]: We want to use TypeScript for deno template

File: deno/sync-with-meilisearch/env.d.ts
Comment (loks0n) [inline deno/sync-with-meilisearch/env.d.ts:1]: Is this file necessary with Deno? Do you get Type errors in your editor without it?

File: deno/sync-with-meilisearch/src/main.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
+import { Client, Databases, Query } from 'https://deno.land/x/appwrite/mod.ts';
+import { getStaticFile, interpolate, throwIfMissing } from './utils.js';
+import { MeiliSearch } from "https://esm.sh/meilisearch";
+import { config as dotEnvConfig } from 'https://deno.land/x/dotenv@v1.0.1/mod.ts';
+
+dotEnvConfig({ export: true });
```
Comment (loks0n) [inline deno/sync-with-meilisearch/src/main.js:6]: We shouldn't need this, ENV vars are set by the runtime

File: deno/sync-with-meilisearch/src/utils.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+const __dirname = path.dirname(__filename);
+const staticFolder = path.join(__dirname, '../static');
+
+/**
+ * Returns the contents of a file in the static folder
+ * @param {string} fileName
+ * @returns {string} Contents of static/{fileName}
+ */
+export function getStaticFile(fileName) {
+  return fs.readFileSync(path.join(staticFolder, fileName)).toString();
```
Comment (loks0n) [inline deno/sync-with-meilisearch/src/utils.js:33]: Is there a more Deno idiomatic way of achieving this? e.g https://docs.deno.com/runtime/tutorials/read_write_files

**Review by loks0n — CHANGES_REQUESTED** [review]

**Comment by Aunali321** [comment]
Almost done with the PR. Have a minor issue where i can run it fine locally and on deno playground but not on appwrite with the same settings.

File: deno/sync-with-meilisearch/src/main.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    log(`Syncing chunk of ${documents.length} documents ...`);
+    await index.addDocuments(documents, { primaryKey: '$id' });
+  } while (cursor !== null);
+
+  log('Sync finished.');
+
+  return new Response('Sync finished.', {status: 200});
+}
+
+serve(handler);
```
Comment (stnguyen90) [inline deno/sync-with-meilisearch/src/main.ts:75]: You should not be running serve here.

File: deno/sync-with-meilisearch/src/main.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
+import { Client, Databases, Query } from "https://deno.land/x/appwrite@9.0.0/mod.ts";
+import { getStaticFile, interpolate, throwIfMissing } from './utils.ts';
+import { MeiliSearch } from "https://esm.sh/meilisearch@0.35.0";
+import { serve } from "https://deno.land/std@0.154.0/http/server.ts";
+
+const handler =  async (req : any, log : any) => {
```
Comment (stnguyen90) [inline deno/sync-with-meilisearch/src/main.ts:6]: This also doesn't match the format required by Appwrite.

**Review by stnguyen90 — CHANGES_REQUESTED** [review]

File: deno/sync-with-meilisearch/src/utils.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+/**
+ * Returns the contents of a file in the static folder
+ */
+export async function getStaticFile(fileName: string): Promise<string> {
+    const staticFolder = "../static/"
+    const filePath = new URL(fileName, `${import.meta.url}/../${staticFolder}`);
+    try {
+        const data = await Deno.readFile(filePath);
+        return new TextDecoder().decode(data);
+    } catch (error) {
```
Comment (loks0n) [inline deno/sync-with-meilisearch/src/utils.ts:25]: I think you can remove this try catch block, since we are just rethrowing the error

File: deno/sync-with-meilisearch/src/main.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  ]);
+
+  if (req.method === 'GET') {
+    const html = interpolate(await getStaticFile('index.html'), {
+      MEILISEARCH_ENDPOINT: Deno.env.get('MEILISEARCH_ENDPOINT'),
+      MEILISEARCH_INDEX_NAME: Deno.env.get('MEILISEARCH_INDEX_NAME'),
+      MEILISEARCH_SEARCH_API_KEY: Deno.env.get('MEILISEARCH_SEARCH_API_KEY'),
+    });
+
+    return new Response(html,{ status: 200, headers: { 'Content-Type': 'text/html; charset=utf-8' } });
```
Comment (loks0n) [inline deno/sync-with-meilisearch/src/main.ts:24]: Can we also format this file with deno fmt?

Comment (loks0n) [inline deno/sync-with-meilisearch/src/main.ts:24]: Response is a deno object, we need to use OPR res object helper methods such as:
`res.send()`, `res.json()`

**Review by loks0n — CHANGES_REQUESTED** [review]

File: deno/sync-with-meilisearch/src/utils.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  }
+  if (missing.length > 0) {
+    throw new Error(`Missing required fields: ${missing.join(', ')}`);
+  }
+}
+
+/**
+ * Returns the contents of a file in the static folder
+ */
+export async function getStaticFile(fileName: string): Promise<string> {
```
Comment (loks0n) [inline deno/sync-with-meilisearch/src/utils.ts:19]: How about this?
```typescript
/*
 * Returns the contents of a file in the static folder
 */
export function getStaticFile(fileName: string): string {
    return Deno.readTextFileSync(
        new URL(import.meta.resolve(`../static/${fileName}`))
    );
}
```

**Comment by loks0n** [comment]
If you still having issues setting up deno runtime on your local instance please join our [discord](appwrite.io/discord) and our awesome community should be able to support!

**Comment by Aunali321** [comment]
Here it is!. sorry for laggy video, it was running on vm. and thanks for coping with me!
[appwrite-hackthon-proof.webm](https://github.com/appwrite/templates/assets/48486084/2e732f77-67bc-4c9f-b905-b84ff49d1a7a)

**Review by loks0n — APPROVED** [review]

File: deno/sync-with-meilisearch/README.md [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+## ⚙️ Configuration
+
+| Setting           | Value         |
+| ----------------- | ------------- |
+| Runtime           | Node (18.0)   |
+| Entrypoint        | `src/main.js` |
+| Build Commands    | `npm install` |
+| Permissions       | `any`         |
+| Timeout (Seconds) | 15            |
```
Comment (stnguyen90) [inline deno/sync-with-meilisearch/README.md:27]: These need to be updated.

**Review by stnguyen90 — CHANGES_REQUESTED** [review]

File: deno/sync-with-meilisearch/README.md [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+**Response**
+
+Sample `204` Response: No content.
+
+## ⚙️ Configuration
+
+| Setting           | Value         |
+| ----------------- | ------------- |
+| Runtime           | Bun (1.35)   |
```
Comment (loks0n) [inline deno/sync-with-meilisearch/README.md:23]: Should be Deno?

File: deno/sync-with-meilisearch/src/main.ts [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    });
+
+    return new res.send(html, 200, {
+      "Content-Type": "text/html; charset=utf-8",
+    });
+  }
+
+  const client = new Client()
+    .setEndpoint(
+      Deno.env.get("APPWRITE_ENDPOINT") ?? "https://cloud.appwrite.io/v1",
```
Comment (loks0n) [inline deno/sync-with-meilisearch/src/main.ts:34]: Is it possible to do Deno.env.get("APPWRITE_ENDPOINT")!! ?

If so could be better.

File: deno/sync-with-meilisearch/src/utils.ts [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+/**
+ * Returns the contents of a file in the static folder
+ */
+export function getStaticFile(fileName: string): string {
+  return Deno.readTextFileSync(
+    new URL(import.meta.resolve(`../static/${fileName}`)),
+  );
+}
+export function interpolate(
```
Comment (loks0n) [inline deno/sync-with-meilisearch/src/utils.ts:24]: Can we add a space and a doc string description here like the other functions?

**Review by loks0n — CHANGES_REQUESTED** [review]

File: deno/sync-with-meilisearch/README.md (hunk 9 above)
Comment (Aunali321) [reply deno/sync-with-meilisearch/README.md:23]: Ah yes. been using Bun alot recently 😅️

File: deno/sync-with-meilisearch/src/main.ts (hunk 10 above)
Comment (Aunali321) [reply deno/sync-with-meilisearch/src/main.ts:34]: it is not possible unfortunately.

**Review by loks0n — APPROVED** [review]

**Comment by Aunali321** [comment]
Can this be merged?

**Comment by gewenyu99** [comment]
Hey there! There were a lot of big PRs during this Hacktoberfest, and we wanted to give everyone ample time to collaborate with our engineering team. If you were able to merge your PRs during October, amazing. If it’s still not merged, don’t worry about it either. Either way, we’ve got your Hacktoberfest swag minted and ready to ship. 

Please comment with your Discord username here so we can contact you about your shipping information to deliver your Hacktoberfest swag.
